From 6b0b9307ff9e9475b6beca031055728cef022cbd Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Mon, 14 Mar 2011 20:22:42 +0200
Subject: [PATCH 1/9] Fix build failures with Borland compiler


commit 58a332df9d82951b5fcc064b568df5f87404f88e
Author: Denis Barbier <bouzim@gmail.com>
Date:   Wed Mar 30 17:22:14 2011 +0200

    Fix a bug introduced in "Fix build failures with Borland compiler"
    
    ViewerTest_ViewerCommands.cxx could not be compiled when WNT is not
    set.  I introduced this bug in 6b0b930 when editing Fotis' changes
    to make them more generic.
    
    This commit must be merged into 6b0b930 before submitting to upstream.

---
 ros/inc/InterfaceGraphic_telem.hxx                 |    4 +
 ros/inc/NCollection_CellFilter.hxx                 |    5 +-
 ros/inc/Standard_CString.hxx                       |    6 +-
 ros/inc/Standard_Macro.hxx                         |    4 +-
 ros/inc/Standard_math.hxx                          |    4 +
 ros/inc/Standard_values.h                          |    2 +-
 ros/inc/math_DoubleTab.lxx                         |    4 +
 ros/src/Draw/CommandWindow.cxx                     |    4 +-
 ros/src/Draw/Draw.cxx                              |    3 +
 ros/src/Draw/Draw_Interpretor.cxx                  |    2 +
 ros/src/Image/Image_PixMap.cxx                     |    6 +
 ros/src/MeshTest/MeshTest_PluginCommands.cxx       |    2 +
 ros/src/NCollection/NCollection_BaseAllocator.cxx  |    2 +
 ros/src/NCollection/NCollection_HeapAllocator.cxx  |    2 +
 ros/src/OSD/OSD.cxx                                |    6 +-
 ros/src/OSD/OSD_Environment.cxx                    |    2 +
 ros/src/OSD/OSD_File.cxx                           |   31 ++--
 ros/src/OSD/OSD_FileNode.cxx                       |   13 +-
 ros/src/OSD/OSD_SystemFont.cxx                     |    1 +
 ros/src/OSD/OSD_WNT.cxx                            |   26 ++--
 ros/src/OSD/OSD_WNT_1.cxx                          |   29 +++-
 ros/src/OSD/OSD_signal_WNT.cxx                     |  163 +++++++++++++++++---
 ros/src/OpenGl/OpenGl_TextRender.cxx               |    3 +
 ros/src/OpenGl/OpenGl_attri.cxx                    |    3 +
 ros/src/OpenGl/OpenGl_graduatedtrihedron.cxx       |    5 +
 ros/src/OpenGl/OpenGl_polygon.cxx                  |    2 +-
 ros/src/OpenGl/OpenGl_togl_print.cxx               |    4 +
 ros/src/OpenGl/OpenGl_txgl.cxx                     |    7 +
 ros/src/PlotMgt/PlotMgt.cxx                        |    2 +-
 ros/src/Poly/Poly_CoherentNode.cxx                 |    2 +
 .../STEPConstruct/STEPConstruct_AP203Context.cxx   |    4 +
 ros/src/Standard/Standard_Macro.hxx                |    6 +-
 ros/src/Standard/Standard_osd_ffi.c                |    2 +-
 ros/src/TObj/TObj_Model.cxx                        |    2 +
 .../Viewer2dTest/Viewer2dTest_DisplayCommands.cxx  |    2 +
 .../Viewer2dTest/Viewer2dTest_ViewerCommands.cxx   |    2 +-
 ros/src/ViewerTest/ViewerTest_ViewerCommands.cxx   |   11 +-
 ros/src/Voxel/Voxel_Reader.cxx                     |    2 +
 ros/src/Voxel/Voxel_Writer.cxx                     |    2 +
 ros/src/VrmlData/VrmlData_Geometry.cxx             |    2 +
 ros/src/VrmlData/VrmlData_Group.cxx                |    2 +
 ros/src/VrmlData/VrmlData_IndexedFaceSet.cxx       |    2 +
 ros/src/VrmlData/VrmlData_Material.cxx             |    2 +
 ros/src/VrmlData/VrmlData_Scene.cxx                |   10 +-
 ros/src/VrmlData/VrmlData_ShapeConvert.cxx         |    8 +-
 ros/src/VrmlData/VrmlData_WorldInfo.cxx            |    2 +
 ros/src/WNT/EHDC.cxx                               |    4 +
 ros/src/WNT/W95_Allocator.cxx                      |    7 +
 ros/src/WNT/WNT_Allocator.cxx                      |    4 +
 ros/src/WNT/WNT_IconBox.cxx                        |    7 +-
 ros/src/WNT/WNT_ImageProcessor.cxx                 |   27 ++++
 ros/src/WNT/WNT_Window.cxx                         |    8 +-
 .../XmlMDataStd/XmlMDataStd_NamedDataDriver.cxx    |    2 +
 ros/src/XmlMXCAFDoc/XmlMXCAFDoc_DimTolDriver.cxx   |    2 +
 54 files changed, 383 insertions(+), 88 deletions(-)

From 7d082abed2b6b7b0132c1344b314d67016e53d2c Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Tue, 22 Mar 2011 07:10:10 +0100
Subject: [PATCH 4/9] Fix LODWORD and HIDWORD macros

Those fixes seem to be generic.
---
 ros/inc/OSD_WNT_1.hxx |    9 +++++++--
 1 files changed, 7 insertions(+), 2 deletions(-)

From c3eecc2595669dcf0c1cf937f83d31f659b3ae30 Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Tue, 22 Mar 2011 07:11:46 +0100
Subject: [PATCH 5/9] BCC uses different symbol names

---
 ros/inc/Standard_Atomic.hxx       |   20 ++++++++++++++------
 ros/inc/Standard_Macro.hxx        |    6 ++++++
 ros/inc/Standard_Real.hxx         |    7 ++++++-
 ros/src/OSD/OSD_SharedLibrary.cxx |   11 ++++++++++-
 4 files changed, 36 insertions(+), 8 deletions(-)

commit 82971f23c27b434cc4a7cfe7cbbf23277d5175ff
Author: Fotis Sioutis <sfotis at gmail dot com>
Date:   Thu Apr 7 16:01:59 2011 +0300

    ftgl fixes
    
    Fixed issues that surfaced when migrated to latest ftgl (2.1.3-rc5)
    with bcc.Also added gl header inclusion.

commit 06b9baa5793862cc1821307e464603508c5817f5
Author: Fotis Sioutis <sfotis at gmail dot com>
Date:   Fri Apr 8 15:05:37 2011 +0300

    ftgl header fixes
    
    Remove conditional gl header inclusion and unprotected ftgl
    header inclusion in main body of code. See :
    http://groups.google.com/group/oce-dev/browse_thread/thread/50ee4ad3336d59de

commit 97e46b7432274807cef47ad9c27936262a1cfe5c
Author: Fotis Sioutis <sfotis at gmail dot com>
Date:   Thu Apr 7 17:08:22 2011 +0300

    Fixed a bcc issue that somehow was not fixed with the bulk fixes at db/ports.

commit 66d637fc8d82a0d6007cb8c93500b4e1d4693943
Author: Denis Barbier <bouzim@gmail.com>
Date:   Mon Mar 14 23:25:17 2011 +0100

    Fix build failures with Mingw
    
    Patch written by Jérôme Robert, downloaded from
     https://github.com/jeromerobert/jCAE/blob/master/occjava/occ-mingw-port.patch

From 70808cfb0129b99c37cea82874ac47740913061b Mon Sep 17 00:00:00 2001
From: Muellni <mmuellen@gmx.de>
Date: Sun, 22 May 2011 19:27:25 +0200
Subject: [PATCH 1/8] fix typo in windows.h include

---
 src/NIS/NIS_View.cxx |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

From b69b768aafd3b9177b9f996369f1a199ebf26dfe Mon Sep 17 00:00:00 2001
From: Muellni <mmuellen@gmx.de>
Date: Sun, 22 May 2011 19:28:24 +0200
Subject: [PATCH 2/8] fix wrong glu.h includes (which fail on non windows
 build hosts)

---
 src/OpenGl/OpenGl_polygon.cxx      |    2 +-
 src/OpenGl/OpenGl_polygonholes.cxx |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

From 10b2ad66de5f245d9ce734c58bc43183e05dc510 Mon Sep 17 00:00:00 2001
From: Muellni <mmuellen@gmx.de>
Date: Sun, 22 May 2011 19:29:18 +0200
Subject: [PATCH 3/8] this fix should only be applied on msvc

---
 inc/W32_Allocator.hxx |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

From 2239f32d9267e0dfdd12e63a62eb91afd19be188 Mon Sep 17 00:00:00 2001
From: Muellni <mmuellen@gmx.de>
Date: Sun, 22 May 2011 19:30:06 +0200
Subject: [PATCH 4/8] fixed missing include for some texture format

---
 src/OpenGl/OpenGl_togl_redraw.cxx |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

From 043e62f6700c8d55d08ab00da54df4f052a46c41 Mon Sep 17 00:00:00 2001
From: Muellni <mmuellen@gmx.de>
Date: Sun, 22 May 2011 19:30:57 +0200
Subject: [PATCH 5/8] ENHMFENUMPROC prototype has been fixed in mingw 3.14

---
 src/WNT/WNT_DDriver.cxx |   11 +++++++++--
 1 files changed, 9 insertions(+), 2 deletions(-)

From b235ac857c94437cb93e1b2b87e66c093e483609 Mon Sep 17 00:00:00 2001
From: Denis Barbier <bouzim@gmail.com>
Date: Mon, 23 May 2011 22:27:56 +0200
Subject: [PATCH 6/8] Fix __try/__finally on Mingw

Apply the same fix as in 66d637f.
---
 src/OpenGl/OpenGl_txgl.cxx |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

From 8a73f5c0b4a8f36a4b7c616dadd6d805ee5c5e52 Mon Sep 17 00:00:00 2001
From: QbProg <tholag@gmail.com>
Date: Tue, 24 May 2011 09:55:42 +0200
Subject: [PATCH 7/8] Made glext.h include specific for mingw

---
 src/OpenGl/OpenGl_togl_redraw.cxx |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

From 8b121e123e9a33388749621817ab2a63cd775e68 Mon Sep 17 00:00:00 2001
From: Denis Barbier <bouzim@gmail.com>
Date: Tue, 24 May 2011 23:35:06 +0200
Subject: [PATCH 8/8] Slightly change again glext.h conditional include

WNT is always defined when compiling with Mingw.
---
 src/OpenGl/OpenGl_togl_redraw.cxx |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit d8d5c7b8b3e4943e7879a9ed547b451b3c3e955f
Author: Denis Barbier <bouzim@gmail.com>
Date:   Thu Apr 7 09:44:58 2011 +0200

    Fix Tk and X11 dependencies in TKDraw and TKViewerTest
    
    Modify src/Draw/Draw_Window.cxx to include tk.h on
    Windows only if _TK is defined.  All Tk instructions
    are protected by _TK on Windows, so this change is safe.
    On Unix machines, tk.h is included unconditionnally.
    
    As all other source files compile without Tk on
    Windows, it is very likely that Tk support on Windows
    is not well tested, so it does not hurt if it is
    disabled.  We can easily reenable it later.
    
    Modify TKViewerTest in a similar manner, Tk is used
    only on Unix, and X11 includes and libraries are added
    since they are used in src/ViewerTest/ViewerTest_ViewerCommands.cxx

From bbfeb103e2136f159ee5d5cbd3f1c5801e04aa50 Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@users.sf.net>
Date: Wed, 19 Oct 2011 10:16:39 +0200
Subject: [PATCH] mingw-w64: IDYES was not declared in this scope
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Fix this compilation error:
src/OSD/OSD_signal_WNT.cxx:194:39: error: ‘IDI_HAND’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:195:32: error: ‘IDYES’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:196:32: error: ‘IDNO’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:197:32: error: ‘IDCANCEL’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx: In function ‘BOOL _osd_ctrl_break_handler(DWORD)’:
src/OSD/OSD_signal_WNT.cxx:608:17: error: ‘MB_ICONEXCLAMATION’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:608:36: error: ‘MessageBeep’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx: In function ‘void _debug_break(Standard_PCharacter)’:
src/OSD/OSD_signal_WNT.cxx:802:39: error: ‘MB_SYSTEMMODAL’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:802:56: error: ‘MB_ICONQUESTION’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:802:74: error: ‘MB_YESNO’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:803:8: error: ‘MessageBox’ was not declared in this scope
src/OSD/OSD_signal_WNT.cxx:803:13: error: ‘IDYES’ was not declared in this scope
---
 src/OSD/OSD_signal_WNT.cxx |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

From 22d95318c17ebbf8c5cbc53ef349f763d976f9a2 Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@users.sf.net>
Date: Wed, 19 Oct 2011 10:43:25 +0200
Subject: [PATCH] mingw-w64: min was not declared in this scope
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

src/OSD/OSD_WNT_1.cxx:481:36: error: ‘min’ was not declared in this scope
---
 src/OSD/OSD_WNT_1.cxx |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

From 297fadf8deca561dc9497a2f6cd311eae21f1d3d Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@users.sf.net>
Date: Thu, 20 Oct 2011 09:35:29 +0200
Subject: [PATCH] mingw-w64: atol was not declared
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

src/OSD/OSD_MAllocHook.cxx:294:24: error: ‘atol’ was not declared in this scope
---
 src/OSD/OSD_MAllocHook.cxx |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

From 2a38a2953e320c721345035008e32798d6500905 Mon Sep 17 00:00:00 2001
From: Jerome Robert <jeromerobert@users.sf.net>
Date: Tue, 25 Oct 2011 14:30:56 +0200
Subject: [PATCH] Define WNT from compiler pre-define macros

---
 inc/Standard_Macro.hxx |    5 ++---
 1 files changed, 2 insertions(+), 3 deletions(-)

From e88ffdd79fee79817f075e36603cda86bd0c69a1 Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Fri, 18 Nov 2011 12:09:35 +0200
Subject: [PATCH] Removal of __BORLANDC__ deprecated macros

---
 src/Draw/CommandWindow.cxx    |    3 ---
 src/OpenGl/OpenGl_FontMgr.cxx |    5 -----
 2 files changed, 0 insertions(+), 8 deletions(-)

From 9cedb2b0eb22262dc81c9a33a62ac406ff12171a Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Fri, 18 Nov 2011 12:13:00 +0200
Subject: [PATCH] stdint.h and stdlib.h inclusion for bcc

While WNT does not include those headers, mingw and
borland needs them.
---
 src/OSD/OSD_MAllocHook.cxx |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

From e26fdfda4b134f1d9053ef8d48a9b7eba0e0db57 Mon Sep 17 00:00:00 2001
From: Fotios Sioutis <sfotis@gmail.com>
Date: Fri, 18 Nov 2011 15:17:56 +0200
Subject: [PATCH] Bcc needs two functions declared as global

This fix has been introduced again in oce in past, but
lost due to upstream changes.Corrects bcc compilation issue.
---
 src/BRepMesh/BRepMesh_CellFilter.hxx |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

Index: oce/src/Draw/Draw.cxx
===================================================================
--- oce.orig/src/Draw/Draw.cxx
+++ oce/src/Draw/Draw.cxx
@@ -45,6 +45,7 @@
 # include <strings.h>
 #endif
 
+#include <stdio.h>
 
 #include <Draw_Window.hxx>
 #include <gp_Pnt2d.hxx>
Index: oce/src/Draw/Draw_Interpretor.cxx
===================================================================
--- oce.orig/src/Draw/Draw_Interpretor.cxx
+++ oce/src/Draw/Draw_Interpretor.cxx
@@ -45,6 +45,8 @@
 #define STDERR_FILENO fileno(stderr)
 #endif
 
+#include <stdio.h>
+
 #if ((TCL_MAJOR_VERSION > 8) || ((TCL_MAJOR_VERSION == 8) && (TCL_MINOR_VERSION >= 1)))
 #define TCL_USES_UTF8
 #endif
Index: oce/src/Draw/Draw_Window.cxx
===================================================================
--- oce.orig/src/Draw/Draw_Window.cxx
+++ oce/src/Draw/Draw_Window.cxx
@@ -1298,7 +1298,9 @@ prompt:
 #include <Draw_Appli.hxx>
 #include <OSD.hxx>
 
-#include <tk.h>
+#ifdef _TK
+# include <tk.h>
+#endif
 
 #define PENWIDTH 1
 #define CLIENTWND 0
Index: oce/src/Font/Font_SystemFont.cxx
===================================================================
--- oce.orig/src/Font/Font_SystemFont.cxx
+++ oce/src/Font/Font_SystemFont.cxx
@@ -19,6 +19,7 @@
 #include <OSD_Path.hxx>
 #include <TCollection_HAsciiString.hxx>
 
+#include <stdio.h>
 
 Font_SystemFont::Font_SystemFont():
 MyFontName(),
Index: oce/src/InterfaceGraphic/InterfaceGraphic_telem.hxx
===================================================================
--- oce.orig/src/InterfaceGraphic/InterfaceGraphic_telem.hxx
+++ oce/src/InterfaceGraphic/InterfaceGraphic_telem.hxx
@@ -20,6 +20,10 @@
 #include <InterfaceGraphic_tgl_all.hxx>
 #include <stdlib.h>
 
+#ifdef __BORLANDC__
+#include <mem.h>
+#endif
+
 struct TEL_TEXTURE_COORD
 {
   float xy[2];
Index: oce/src/MeshTest/MeshTest_PluginCommands.cxx
===================================================================
--- oce.orig/src/MeshTest/MeshTest_PluginCommands.cxx
+++ oce/src/MeshTest/MeshTest_PluginCommands.cxx
@@ -47,6 +47,8 @@
 #include <BRepMesh_Edge.hxx>
 #include <NCollection_Map.hxx>
 
+#include <stdio.h>
+
 static Standard_Integer mpnames           (Draw_Interpretor& , Standard_Integer , const char** );
 static Standard_Integer mpsetdefaultname  (Draw_Interpretor& , Standard_Integer , const char** );
 static Standard_Integer mpgetdefaultname  (Draw_Interpretor& , Standard_Integer , const char** );
Index: oce/src/NCollection/NCollection_CellFilter.hxx
===================================================================
--- oce.orig/src/NCollection/NCollection_CellFilter.hxx
+++ oce/src/NCollection/NCollection_CellFilter.hxx
@@ -228,8 +228,9 @@ public:
     iterateInspect (Inspector::Dimension-1, aCell, 
                     aCellMin, aCellMax, theInspector);
   }
-
-#if defined(__SUNPRO_CC) && (__SUNPRO_CC <= 0x530)
+//Borland needs to have the Cell struct public in order to access it from
+//the global functions HashCode and IsEqual ... See below on the declarations of them
+#if defined(__SUNPRO_CC) && (__SUNPRO_CC <= 0x530) || defined(__BORLANDC__)
 public: // work-around against obsolete SUN WorkShop 5.3 compiler
 #else
 protected:
Index: oce/src/OSD/OSD.cxx
===================================================================
--- oce.orig/src/OSD/OSD.cxx
+++ oce/src/OSD/OSD.cxx
@@ -72,7 +72,7 @@ Standard_Boolean OSD::CStringToReal(cons
 //=======================================================================
 
 #ifdef WNT
-# include <Windows.h>
+# include <windows.h>
 #if !defined(__CYGWIN32__) && !defined(__MINGW32__)
 //# include <Mapiwin.h>
 #endif
Index: oce/src/OSD/OSD_Environment.cxx
===================================================================
--- oce.orig/src/OSD/OSD_Environment.cxx
+++ oce/src/OSD/OSD_Environment.cxx
@@ -238,6 +238,8 @@ Standard_Integer OSD_Environment::Error(
 
 #include <windows.h>
 
+#include <stdio.h>
+
 #ifdef _MSC_VER
 #pragma warning( disable : 4700 )
 #endif
Index: oce/src/OSD/OSD_FileNode.cxx
===================================================================
--- oce.orig/src/OSD/OSD_FileNode.cxx
+++ oce/src/OSD/OSD_FileNode.cxx
@@ -445,6 +445,8 @@ Standard_Integer OSD_FileNode::Error()co
 # include <tchar.h>
 #endif  // _INC_TCHAR
 
+#include <stdio.h>
+
 #define TEST_RAISE( arg ) _test_raise (  fName, ( arg )  )
 #define RAISE( arg ) Standard_ProgramError :: Raise (  ( arg )  )
 
Index: oce/src/OSD/OSD_MAllocHook.cxx
===================================================================
--- oce.orig/src/OSD/OSD_MAllocHook.cxx
+++ oce/src/OSD/OSD_MAllocHook.cxx
@@ -15,7 +15,7 @@
 
 #include <OSD_MAllocHook.hxx>
 
-#ifndef WNT
+#if !defined(WNT) || defined(__MINGW32__) || defined(__BORLANDC__)
 #if !defined __STDC_LIMIT_MACROS
 #define __STDC_LIMIT_MACROS
 #endif
Index: oce/src/OSD/OSD_MemInfo.cxx
===================================================================
--- oce.orig/src/OSD/OSD_MemInfo.cxx
+++ oce/src/OSD/OSD_MemInfo.cxx
@@ -18,8 +18,8 @@
   #include <winbase.h>
   #include <process.h>
   #include <malloc.h>
-  #include <Psapi.h>
-  #ifdef _MSC_VER
+  #include <psapi.h>
+  #if defined(_MSC_VER) || defined(__BORLANDC__)
     #pragma comment(lib, "Psapi.lib")
   #endif
 #elif (defined(__APPLE__))
@@ -35,6 +35,18 @@
 #include <sstream>
 #include <fstream>
 
+#ifdef HAVE_CONFIG_H
+# include <oce-config.h>
+#endif
+
+#ifdef HAVE_UNISTD_H
+# include <unistd.h>
+#endif
+
+#ifdef HAVE_SYS_TYPES_H
+# include <sys/types.h>
+#endif
+
 #include <OSD_MemInfo.hxx>
 
 // =======================================================================
Index: oce/src/OSD/OSD_Process.cxx
===================================================================
--- oce.orig/src/OSD/OSD_Process.cxx
+++ oce/src/OSD/OSD_Process.cxx
@@ -230,7 +230,7 @@ Standard_Integer OSD_Process::Error()con
 #include <Quantity_Date.hxx>
 
 #include <OSD_WNT_1.hxx>
-#include <LMCONS.H> /// pour UNLEN  ( see MSDN about GetUserName() )
+#include <lmcons.h> /// pour UNLEN  ( see MSDN about GetUserName() )
 
 #ifdef _MSC_VER
 #pragma warning( disable : 4700 )
Index: oce/src/OSD/OSD_SharedLibrary.cxx
===================================================================
--- oce.orig/src/OSD/OSD_SharedLibrary.cxx
+++ oce/src/OSD/OSD_SharedLibrary.cxx
@@ -316,9 +316,18 @@ OSD_Function OSD_SharedLibrary :: DlSymb
 
  OSD_Function func = ( OSD_Function )GetProcAddress (  ( HMODULE )myHandle, Name  );
 
+ if ( func == NULL ) {
+#ifdef __BORLANDC__ //borland decorates exported C style routines with '_'
+  Standard_Character* myAlternativeName = new Standard_Character[ strlen ( Name ) + 1 + 1 ];
+  static Standard_Character* myAddition = "_";
+  strcpy(myAlternativeName, myAddition);
+  strcat(myAlternativeName, Name);
+  func = ( OSD_Function )GetProcAddress (  ( HMODULE )myHandle, myAlternativeName  );
+  delete [] myAlternativeName;
  if ( func == NULL )
-
+#endif
   lastDLLError = GetLastError ();
+ }
 
  return func;
 
Index: oce/src/OSD/OSD_WNT_1.hxx
===================================================================
--- oce.orig/src/OSD/OSD_WNT_1.hxx
+++ oce/src/OSD/OSD_WNT_1.hxx
@@ -65,8 +65,13 @@ enum MB_ITEMTYPE { MBT_BUTTON, MBT_ICON
 //#define LODWORD( a ) ( DWORD )(   ( ( DWORDLONG )( a ) ) & 0x00000000FFFFFFFF   )
 //#define HIDWORD( a ) ( DWORD )(   ( ( DWORDLONG )( a ) ) >> 32                  )
 
-#define LODWORD( a ) ( DWORD )(   ( ( _int64 )( a ) ) & 0x00000000FFFFFFFF   )
-#define HIDWORD( a ) ( DWORD )(   ( ( _int64 )( a ) ) >> 32                  )
+#ifdef __BORLANDC__
+# define LODWORD( a ) ( DWORD )(   ( ( DWORDLONG )( a ) ) & 0x00000000FFFFFFFF   )
+# define HIDWORD( a ) ( DWORD )(   ( ( DWORDLONG )( a ) ) >> 32                  )
+#else
+# define LODWORD( a ) ( DWORD )(   ( ( _int64 )( a ) ) & 0x00000000FFFFFFFF   )
+# define HIDWORD( a ) ( DWORD )(   ( ( _int64 )( a ) ) >> 32                  )
+#endif
 
 typedef struct _group_sid {
 
Index: oce/src/OSD/OSD_signal_WNT.cxx
===================================================================
--- oce.orig/src/OSD/OSD_signal_WNT.cxx
+++ oce/src/OSD/OSD_signal_WNT.cxx
@@ -21,6 +21,12 @@
 #ifdef NOUSER
 #undef NOUSER
 #endif
+#ifdef NOICONS
+#undef NOICONS
+#endif
+#ifdef NOGDI
+#undef NOGDI
+#endif
 #include <windows.h>
 
 #include <OSD_Exception_ACCESS_VIOLATION.hxx>
@@ -50,6 +56,136 @@
 #include <malloc.h>
 #endif
 
+#ifdef __BORLANDC__
+# include <malloc.h>
+# define FPE_DENORMAL 0x82
+
+# define _EM_INVALID EM_INVALID
+# define _EM_DENORMAL EM_DENORMAL
+# define _EM_ZERODIVIDE EM_ZERODIVIDE
+# define _EM_OVERFLOW EM_OVERFLOW
+
+# define _FPE_INVALID FPE_INVALID
+# define _FPE_DENORMAL FPE_DENORMAL
+# define _FPE_ZERODIVIDE FPE_ZERODIVIDE
+# define _FPE_OVERFLOW FPE_OVERFLOW
+# define _FPE_UNDERFLOW   FPE_UNDERFLOW
+# define _FPE_INEXACT  FPE_INEXACT
+
+DWORD PlatformId(void)
+{
+  OSVERSIONINFO ovi;
+  ovi.dwOSVersionInfoSize = sizeof(ovi);
+  GetVersionEx(&ovi);
+  return ovi.dwPlatformId;
+}
+
+/* These magic numbers are from the MS header files */
+# define MIN_STACK_WIN9X 17
+# define MIN_STACK_WINNT 2
+
+/*
+ * This function does the same thing as _resetstkoflw(), which is only
+ * available in DevStudio .net and later.
+ * Returns 0 for failure, 1 for success.
+ */
+int _resetstkoflw(void)
+{
+  BYTE  *pStackPtr;
+  BYTE  *pGuardPage;
+  BYTE  *pStackBase;
+  BYTE  *pLowestPossiblePage;
+  MEMORY_BASIC_INFORMATION mbi;
+  SYSTEM_INFO si;
+  DWORD  nPageSize;
+  DWORD  dummy;
+
+  DWORD  g_PlatformId = PlatformId();
+
+  /* This code will not work on win32s. */
+  if (g_PlatformId == VER_PLATFORM_WIN32s)
+    return 0;
+
+  /* We need to know the system page size. */
+  GetSystemInfo(&si);
+  nPageSize = si.dwPageSize;
+
+  /* ...and the current stack pointer */
+  pStackPtr = (BYTE*)_alloca(1);
+
+  /* ...and the base of the stack. */
+  if (VirtualQuery(pStackPtr, &mbi, sizeof mbi) == 0)
+    return 0;
+  pStackBase = (BYTE*)mbi.AllocationBase;
+
+  /* ...and the page thats min_stack_req pages away from stack base; this is
+   * the lowest page we could use. */
+  pLowestPossiblePage = pStackBase + ((g_PlatformId == VER_PLATFORM_WIN32_NT)
+               ? MIN_STACK_WINNT : MIN_STACK_WIN9X) * nPageSize;
+
+  /* On Win95, we want the next page down from the end of the stack. */
+  if (g_PlatformId == VER_PLATFORM_WIN32_WINDOWS) {
+    /* Find the page that's only 1 page down from the page that the stack
+     * ptr is in. */
+    pGuardPage = (BYTE*)((DWORD)nPageSize * (((DWORD)pStackPtr
+                            / (DWORD)nPageSize) - 1));
+    if (pGuardPage < pLowestPossiblePage)
+      return 0;
+
+    /* Apply the noaccess attribute to the page -- there's no guard
+     * attribute in win95-type OSes. */
+    if (!VirtualProtect(pGuardPage, nPageSize, PAGE_NOACCESS, &dummy))
+      return 0;
+  }
+  else {
+    /* On NT, however, we want the first committed page in the stack Start
+     * at the stack base and move forward through memory until we find a
+     * committed block. */
+    BYTE *pBlock = pStackBase;
+
+    for (;;) {
+      if (VirtualQuery(pBlock, &mbi, sizeof mbi) == 0)
+        return 0;
+
+      pBlock += mbi.RegionSize;
+
+      if (mbi.State & MEM_COMMIT)
+        break;
+    }
+
+    /* mbi now describes the first committed block in the stack. */
+    if (mbi.Protect & PAGE_GUARD)
+      return 1;
+
+    /* decide where the guard page should start */
+    if ((mbi.BaseAddress) < pLowestPossiblePage)
+      pGuardPage = pLowestPossiblePage;
+    else
+      pGuardPage = (BYTE*)mbi.BaseAddress;
+
+    /* allocate the guard page */
+    if (!VirtualAlloc(pGuardPage, nPageSize, MEM_COMMIT, PAGE_READWRITE))
+      return 0;
+
+    /* apply the guard attribute to the page */
+    if (!VirtualProtect(pGuardPage, nPageSize, PAGE_READWRITE | PAGE_GUARD,
+                                      &dummy))
+      return 0;
+  }
+
+  return 1;
+}
+
+typedef void (*_se_translator_function)(unsigned int, EXCEPTION_POINTERS*);
+
+_CRTIMP _se_translator_function __cdecl _set_se_translator(_se_translator_function _NewPtFunc)
+{
+  return _NewPtFunc;
+}
+
+#endif
+
+
 #include <process.h>
 #include <signal.h>
 #include <float.h>
@@ -198,7 +334,7 @@ static LONG CallHandler (DWORD dwExcepti
 
   case EXCEPTION_STACK_OVERFLOW:
 //      cout << "CallHandler : EXCEPTION_STACK_OVERFLOW:" << endl ;
-#if defined( _MSC_VER ) && ( _MSC_VER >= 1300 )
+#if (defined( _MSC_VER ) && ( _MSC_VER >= 1300 )) || defined(__BORLANDC__)
     // try recovering from stack overflow: available in MS VC++ 7.0
     if (!_resetstkoflw())
       lstrcpy (  buffer, TEXT( "Unrecoverable STACK OVERFLOW" )  );
@@ -308,7 +444,7 @@ static void SIGWntHandler (int signum, i
 //purpose  : Translate Structural Exceptions into C++ exceptions
 //           Will be used when user's code is compiled with /EHa option
 //=======================================================================
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__BORLANDC__)
 
 // If this file compiled with the default MSVC options for exception
 // handling (/GX or /EHsc) then the following warning is issued:
@@ -383,7 +519,7 @@ void OSD::SetSignal (const Standard_Bool
   fCtrlBrk = Standard_False;
   SetConsoleCtrlHandler (&_osd_ctrl_break_handler, TRUE);
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__BORLANDC__)
 //  _se_translator_function pOldSeFunc = 
     _set_se_translator (TranslateSE);
 #endif
@@ -489,8 +625,10 @@ static LONG __fastcall _osd_raise ( DWOR
     case EXCEPTION_FLT_INVALID_OPERATION:
     case EXCEPTION_FLT_DENORMAL_OPERAND:
     case EXCEPTION_FLT_INEXACT_RESULT:
+#if defined(_MSC_VER) || defined(__BORLANDC__)
     case STATUS_FLOAT_MULTIPLE_TRAPS:
     case STATUS_FLOAT_MULTIPLE_FAULTS:
+#endif
       Standard_NumericError::Raise (msg);
       break;
     default:
Index: oce/src/OpenGl/OpenGl_GraduatedTrihedron.cxx
===================================================================
--- oce.orig/src/OpenGl/OpenGl_GraduatedTrihedron.cxx
+++ oce/src/OpenGl/OpenGl_GraduatedTrihedron.cxx
@@ -130,6 +130,11 @@ static char getFarestCorner(float d000,
     return 8; /* d111 */
 }
 
+#ifdef __BORLANDC__
+# define sqrtf std::sqrtf
+# define fabsf (float)fabs 
+#endif
+
 static void drawArrow(float x1, float y1, float z1,
                       float x2, float y2, float z2,
                       float xn, float yn, float zn)
Index: oce/src/Poly/Poly_CoherentNode.cxx
===================================================================
--- oce.orig/src/Poly/Poly_CoherentNode.cxx
+++ oce/src/Poly/Poly_CoherentNode.cxx
@@ -20,6 +20,8 @@
 #pragma warning(disable:4996)
 #endif
 
+#include <stdio.h>
+
 //=======================================================================
 //function : Clear
 //purpose  : 
Index: oce/src/STEPConstruct/STEPConstruct_AP203Context.cxx
===================================================================
--- oce.orig/src/STEPConstruct/STEPConstruct_AP203Context.cxx
+++ oce/src/STEPConstruct/STEPConstruct_AP203Context.cxx
@@ -59,6 +59,10 @@
 #include <StepAP203_HArray1OfApprovedItem.hxx>
 #include <StepBasic_ProductCategory.hxx>
 
+#ifdef __BORLANDC__
+# define timezone _timezone
+#endif
+
 //=======================================================================
 //function : STEPConstruct_AP203Context
 //purpose  : 
Index: oce/src/Standard/Handle_Standard_Transient.hxx
===================================================================
--- oce.orig/src/Standard/Handle_Standard_Transient.hxx
+++ oce/src/Standard/Handle_Standard_Transient.hxx
@@ -30,7 +30,7 @@
 #include <Standard_Transient_proto.hxx>
 #endif
 
-#ifdef _WIN32
+#ifdef _MSC_VER
 // Disable the warning "conversion from 'unsigned int' to Standard_Transient *"
 #pragma warning (push)
 #pragma warning (disable:4312)
@@ -242,7 +242,7 @@ inline Standard_Boolean IsEqual (const H
   return theFirst == theSecond;
 }
 
-#ifdef _WIN32
+#ifdef _MSC_VER
 #pragma warning (pop)
 #endif
 
Index: oce/src/Standard/Standard_Atomic.hxx
===================================================================
--- oce.orig/src/Standard/Standard_Atomic.hxx
+++ oce/src/Standard/Standard_Atomic.hxx
@@ -55,11 +55,18 @@ int Standard_Atomic_Decrement (volatile
   return __sync_sub_and_fetch (theValue, 1);
 }
 
-#elif defined(_WIN32)
+#elif defined(_MSC_VER) || defined(__BORLANDC__)
+#ifdef __BORLANDC__
+extern "C" {
+  __declspec(dllimport) long __stdcall InterlockedIncrement ( long volatile *lpAddend);
+  __declspec(dllimport) long __stdcall InterlockedDecrement ( long volatile *lpAddend);
+ }
+#else
 extern "C" {
   long _InterlockedIncrement (volatile long* lpAddend);
   long _InterlockedDecrement (volatile long* lpAddend);
 }
+#endif
 
 #if defined(_MSC_VER)
   // force intrinsic instead of WinAPI calls
Index: oce/src/Standard/Standard_CString.hxx
===================================================================
--- oce.orig/src/Standard/Standard_CString.hxx
+++ oce/src/Standard/Standard_CString.hxx
@@ -29,6 +29,8 @@
 
 # ifdef _MSC_VER
 #  define strcasecmp _stricmp
+#  elif defined(__BORLANDC__)
+#    define strcasecmp stricmp
 # endif
 
 # include <string.h>
Index: oce/src/Standard/Standard_MMgrOpt.cxx
===================================================================
--- oce.orig/src/Standard/Standard_MMgrOpt.cxx
+++ oce/src/Standard/Standard_MMgrOpt.cxx
@@ -852,7 +852,11 @@ void Standard_MMgrOpt::FreeMemory (Stand
     // recover handle to the memory mapping stored just before the block
     const HANDLE * aMBlock = (const HANDLE *)aBlock;
     HANDLE hMap = *(--aMBlock);
+#ifndef __MINGW32__
     UnmapViewOfFile((LPCVOID)aMBlock);
+#else
+    UnmapViewOfFile(const_cast<void*>((LPCVOID)aMBlock));
+#endif
     CloseHandle (hMap);
 #endif
   }
Index: oce/src/Standard/Standard_Macro.hxx
===================================================================
--- oce.orig/src/Standard/Standard_Macro.hxx
+++ oce/src/Standard/Standard_Macro.hxx
@@ -29,9 +29,8 @@
 // Windows-specific definitions
 //======================================================
 
-// check if WNT macro is not defined but compiler is MSVC
-#if defined(_MSC_VER) && !defined(WNT)
-#error "Wrong compiler options has been detected. Add /DWNT option for proper compilation!!!!!"
+#if (defined(_MSC_VER) || defined(__BORLANDC__) || defined(__MINGW32__) || defined(__MINGW64__)) && !defined(WNT)
+#define WNT
 #endif
 
 # if defined(WNT) && !defined(HAVE_NO_DLL)
@@ -60,7 +59,7 @@
 #define WIN32_LEAN_AND_MEAN   /* exclude extra Windows stuff */
 #endif
 #ifndef NOMINMAX
-#define NOMINMAX NOMINMAX     /* avoid #define min() and max() */
+#define NOMINMAX              /* avoid #define min() and max() */
 #endif
 #ifndef NOMSG
 #define NOMSG NOMSG           /* avoid #define SendMessage etc. */
@@ -130,9 +129,15 @@
 #   if !defined(WNT) || defined(__Standard_DLL) || defined(__FSD_DLL) || defined(__MMgt_DLL) || defined(__OSD_DLL) || defined(__Plugin_DLL) || defined(__Quantity_DLL) || defined(__Resource_DLL) || defined(__SortTools_DLL) || defined(__StdFail_DLL) || defined(__Storage_DLL) || defined(__TColStd_DLL) || defined(__TCollection_DLL) || defined(__TShort_DLL) || defined(__Units_DLL) || defined(__UnitsAPI_DLL) || defined(__Dico_DLL) || defined(__Message_DLL)
 #    define __Standard_API Standard_EXPORT
 #    define __Standard_APIEXTERN Standard_EXPORTEXTERN
+#      ifdef __BORLANDC__
+#      define __Standard_APIEXTERNC Standard_EXPORTEXTERNC
+#      endif
 #   else
 #    define __Standard_API Standard_IMPORT
 #    define __Standard_APIEXTERN Standard_IMPORT
+#      ifdef __BORLANDC__
+#      define __Standard_APIEXTERNC Standard_IMPORTC
+#      endif
 #   endif  // __Standard_DLL
 //#  else
 //#   define __Standard_API
Index: oce/src/Standard/Standard_values.h
===================================================================
--- oce.orig/src/Standard/Standard_values.h
+++ oce/src/Standard/Standard_values.h
@@ -17,7 +17,7 @@
 #ifndef _Standard_values_HeaderFile
 # define _Standard_values_HeaderFile
 
-#if defined(_MSC_VER) && !defined(WNT)
+#if (defined(_MSC_VER) || defined(__BORLANDC__)) && !defined(WNT)
 #error "Wrong compiler options has been detected. Add /DWNT option for proper compilation!!!!!"
 #endif
 
@@ -40,6 +40,10 @@
 # include <limits>
 #endif
 
+#ifdef __MINGW32__
+# include <limits.h>
+#endif
+
 #if defined (__hpux) || defined (HPUX)
 # ifdef MAXINT
 #  undef MAXINT
Index: oce/src/TObj/TObj_Model.cxx
===================================================================
--- oce.orig/src/TObj/TObj_Model.cxx
+++ oce/src/TObj/TObj_Model.cxx
@@ -48,6 +48,8 @@
   #include <unistd.h>
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE(TObj_Model,MMgt_TShared)
 IMPLEMENT_STANDARD_RTTIEXT(TObj_Model,MMgt_TShared)
 
Index: oce/src/TestTopOpe/TestTopOpe_BOOPCommands.cxx
===================================================================
--- oce.orig/src/TestTopOpe/TestTopOpe_BOOPCommands.cxx
+++ oce/src/TestTopOpe/TestTopOpe_BOOPCommands.cxx
@@ -322,7 +322,11 @@ Standard_Integer BOOPCHK(Draw_Interpreto
 {
   if (!strcmp(a[0],"tchk")) { 
 //    Standard_IMPORT extern Standard_Integer TopOpeBRepTool_BOOOPE_CHECK_DEB;
+#ifdef __BORLANDC__
+    extern Standard_Integer TopOpeBRepTool_BOOOPE_CHECK_DEB;
+#else
     Standard_IMPORT Standard_Integer TopOpeBRepTool_BOOOPE_CHECK_DEB;
+#endif
     if ( na >= 2 ) TopOpeBRepTool_BOOOPE_CHECK_DEB = (Draw::Atoi(a[1])) ? 1 : 0;
     return 0;
   }
Index: oce/src/ViewerTest/ViewerTest_ViewerCommands.cxx
===================================================================
--- oce.orig/src/ViewerTest/ViewerTest_ViewerCommands.cxx
+++ oce/src/ViewerTest/ViewerTest_ViewerCommands.cxx
@@ -105,6 +105,12 @@ static const char THE_KEY_DELETE = 127;
 //==============================================================================
 //  VIEWER GLOBAL VARIABLES
 //==============================================================================
+#ifndef max
+# define max(a, b)  (((a) > (b)) ? (a) : (b))
+#endif
+#ifndef min
+# define min(a, b)  (((a) < (b)) ? (a) : (b)) 
+#endif
 
 Standard_IMPORT Standard_Boolean Draw_VirtualWindows;
 Standard_IMPORT Standard_Boolean Draw_Interprete (const char* theCommand);
@@ -2129,22 +2135,6 @@ int ViewerMainLoop(Standard_Integer argc
 
 #elif !defined(__APPLE__) || defined(MACOSX_USE_GLX)
 
-int min( int a, int b )
-{
-  if( a<b )
-    return a;
-  else
-    return b;
-}
-
-int max( int a, int b )
-{
-  if( a>b )
-    return a;
-  else
-    return b;
-}
-
 int ViewerMainLoop(Standard_Integer argc, const char** argv)
 
 {
Index: oce/src/Voxel/Voxel_Reader.cxx
===================================================================
--- oce.orig/src/Voxel/Voxel_Reader.cxx
+++ oce/src/Voxel/Voxel_Reader.cxx
@@ -22,6 +22,8 @@
 
 #include <TCollection_AsciiString.hxx>
 
+#include <stdio.h>
+
 #define CATCH_FREAD_1_ERR(stmt) if (1 != (stmt)) { fclose(f); return Standard_False; }
 #define CATCH_FGETS_ERR(stmt) if (NULL == (stmt)) { fclose(f); return Standard_False; }
 
Index: oce/src/Voxel/Voxel_Writer.cxx
===================================================================
--- oce.orig/src/Voxel/Voxel_Writer.cxx
+++ oce/src/Voxel/Voxel_Writer.cxx
@@ -19,6 +19,8 @@
 #include <Precision.hxx>
 #include <TCollection_AsciiString.hxx>
 
+#include <stdio.h>
+
 Voxel_Writer::Voxel_Writer():myFormat(Voxel_VFF_ASCII),myBoolVoxels(0),myColorVoxels(0),myFloatVoxels(0)
 {
 
Index: oce/src/VrmlData/VrmlData_Geometry.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_Geometry.cxx
+++ oce/src/VrmlData/VrmlData_Geometry.cxx
@@ -37,6 +37,8 @@
 #pragma warning (disable:4996)
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_Geometry, VrmlData_Node)
 IMPLEMENT_STANDARD_RTTIEXT (VrmlData_Geometry, VrmlData_Node)
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_Box, VrmlData_Geometry)
Index: oce/src/VrmlData/VrmlData_Group.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_Group.cxx
+++ oce/src/VrmlData/VrmlData_Group.cxx
@@ -28,6 +28,8 @@
 #pragma warning (disable:4996)
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_Group, VrmlData_Node)
 IMPLEMENT_STANDARD_RTTIEXT (VrmlData_Group, VrmlData_Node)
 
Index: oce/src/VrmlData/VrmlData_IndexedFaceSet.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_IndexedFaceSet.cxx
+++ oce/src/VrmlData/VrmlData_IndexedFaceSet.cxx
@@ -34,6 +34,8 @@
 #pragma warning (disable:4996)
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_Faceted, VrmlData_Geometry)
 IMPLEMENT_STANDARD_RTTIEXT (VrmlData_Faceted, VrmlData_Geometry)
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_IndexedFaceSet, VrmlData_Faceted)
Index: oce/src/VrmlData/VrmlData_Material.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_Material.cxx
+++ oce/src/VrmlData/VrmlData_Material.cxx
@@ -24,6 +24,8 @@
 #pragma warning (disable:4996)
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_Material, VrmlData_Node)
 IMPLEMENT_STANDARD_RTTIEXT (VrmlData_Material, VrmlData_Node)
 
Index: oce/src/VrmlData/VrmlData_Scene.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_Scene.cxx
+++ oce/src/VrmlData/VrmlData_Scene.cxx
@@ -92,8 +92,12 @@ const Handle(VrmlData_Node)& VrmlData_Sc
   if (theN.IsNull() == Standard_False)
     if (theN->IsKind (STANDARD_TYPE(VrmlData_WorldInfo)) == Standard_False) {
       myMutex.Lock();
-      const Handle(VrmlData_Node)& aNode =
-        myAllNodes.Append ((&theN->Scene()== this) ? theN : theN->Clone (NULL));
+      Handle(VrmlData_Node) theItem;
+      if (&theN->Scene() == this)
+        theItem = theN;
+      else
+        theItem = theN->Clone (NULL);
+      const Handle(VrmlData_Node)& aNode = myAllNodes.Append (theItem);
       // Name is checked for uniqueness. If not, letter 'D' is appended until
       // the name proves to be unique.
       if (aNode->Name()[0] != '\0')
Index: oce/src/VrmlData/VrmlData_ShapeConvert.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_ShapeConvert.cxx
+++ oce/src/VrmlData/VrmlData_ShapeConvert.cxx
@@ -295,8 +295,12 @@ void VrmlData_ShapeConvert::Convert (con
         if (aTShapeNode.IsNull() == Standard_False) {
           const Handle(VrmlData_ShapeNode) aShapeNode =
             new VrmlData_ShapeNode (myScene, 0L);
-          aShapeNode->SetAppearance (ShapeType[i] == TopAbs_FACE ?
-                                     defaultMaterialFace():defaultMaterialEdge());
+          Handle(VrmlData_Appearance) theApp;
+          if (ShapeType[i] == TopAbs_FACE)
+            theApp = defaultMaterialFace();
+          else
+            theApp = defaultMaterialEdge();
+          aShapeNode->SetAppearance (theApp);
           myScene.AddNode (aShapeNode, Standard_False);
           aShapeNode->SetGeometry (aTShapeNode);
           if (aLoc.IsIdentity())
Index: oce/src/VrmlData/VrmlData_WorldInfo.cxx
===================================================================
--- oce.orig/src/VrmlData/VrmlData_WorldInfo.cxx
+++ oce/src/VrmlData/VrmlData_WorldInfo.cxx
@@ -22,6 +22,8 @@
 #pragma warning (disable:4996)
 #endif
 
+#include <stdio.h>
+
 IMPLEMENT_STANDARD_HANDLE  (VrmlData_WorldInfo, VrmlData_Node)
 IMPLEMENT_STANDARD_RTTIEXT (VrmlData_WorldInfo, VrmlData_Node)
 
Index: oce/src/WNT/WNT_WClass.cxx
===================================================================
--- oce.orig/src/WNT/WNT_WClass.cxx
+++ oce/src/WNT/WNT_WClass.cxx
@@ -59,7 +59,7 @@ WNT_WClass::WNT_WClass (
 
   lpszName = new char[ strlen ( aClassName ) + 1 ];
   strcpy ( (Standard_PCharacter)lpszName, aClassName );
-  lpfnWndProc = wc.lpfnWndProc;
+  lpfnWndProc = (void*)wc.lpfnWndProc;
 
 }  // end constructor
 
Index: oce/src/WNT/WNT_Window.cxx
===================================================================
--- oce.orig/src/WNT/WNT_Window.cxx
+++ oce/src/WNT/WNT_Window.cxx
@@ -221,10 +221,10 @@ Aspect_TypeOfResize WNT_Window :: DoResi
 
   if (wp.showCmd != SW_SHOWMINIMIZED)
   {
-    if (Abs (wp.rcNormalPosition.left   - aXLeft  ) > 2) mask |= 1;
-    if (Abs (wp.rcNormalPosition.right  - aXRight ) > 2) mask |= 2;
-    if (Abs (wp.rcNormalPosition.top    - aYTop   ) > 2) mask |= 4;
-    if (Abs (wp.rcNormalPosition.bottom - aYBottom) > 2) mask |= 8;
+    if (Abs ((Standard_Integer) wp.rcNormalPosition.left   - aXLeft  ) > 2) mask |= 1;
+    if (Abs ((Standard_Integer) wp.rcNormalPosition.right  - aXRight ) > 2) mask |= 2;
+    if (Abs ((Standard_Integer) wp.rcNormalPosition.top    - aYTop   ) > 2) mask |= 4;
+    if (Abs ((Standard_Integer) wp.rcNormalPosition.bottom - aYBottom) > 2) mask |= 8;
 
     switch (mask)
     {
Index: oce/src/XmlMDataStd/XmlMDataStd_NamedDataDriver.cxx
===================================================================
--- oce.orig/src/XmlMDataStd/XmlMDataStd_NamedDataDriver.cxx
+++ oce/src/XmlMDataStd/XmlMDataStd_NamedDataDriver.cxx
@@ -51,6 +51,8 @@ IMPLEMENT_DOMSTRING (LastRealArrIndex,
 IMPLEMENT_DOMSTRING (ExtString, "string")
 IMPLEMENT_DOMSTRING (Value,     "value")
 
+#include <stdio.h>
+
 #define OCC6010
 
 #ifdef WNT
Index: oce/src/XmlMXCAFDoc/XmlMXCAFDoc_DimTolDriver.cxx
===================================================================
--- oce.orig/src/XmlMXCAFDoc/XmlMXCAFDoc_DimTolDriver.cxx
+++ oce/src/XmlMXCAFDoc/XmlMXCAFDoc_DimTolDriver.cxx
@@ -19,6 +19,8 @@
 #include <TCollection_HAsciiString.hxx>
 #include <TColStd_HArray1OfReal.hxx>
 
+#include <stdio.h>
+
 IMPLEMENT_DOMSTRING (NameIndexString,  "name")
 IMPLEMENT_DOMSTRING (DescrIndexString, "descr")
 IMPLEMENT_DOMSTRING (FirstIndexString, "first")
Index: oce/src/math/math_DoubleTab.lxx
===================================================================
--- oce.orig/src/math/math_DoubleTab.lxx
+++ oce/src/math/math_DoubleTab.lxx
@@ -17,6 +17,10 @@
 #include <Standard_OutOfRange.hxx>
 #include <cstring>
 
+#ifdef __BORLANDC__
+#include <mem.h>
+#endif
+
 inline Standard_Real& math_DoubleTab::Value (const Standard_Integer RowIndex,
 				    const Standard_Integer ColIndex) const
 {
